Fix warnings about implicit declaration of memcpy

cc @eunchan @a-will 

These warnings about an implicit declaration of memcpy are probably related to the manual name mangling from #14009. When compiling for the host machine, the OpenTitan memory functions' names are prefixed with `ot_` to disambiguate from libc's declarations and definitions.

```console
% ./bazelisk.sh build //sw/device/lib/base:mmio_unittest
INFO: Analyzed target //sw/device/lib/base:mmio_unittest (1 packages loaded, 265 targets configured).
INFO: Found 1 target...
INFO: From Compiling sw/device/lib/base/mmio.c:
sw/device/lib/base/mmio.c: In function 'mmio_region_memcpy32':
sw/device/lib/base/mmio.c:55:7: warning: implicit declaration of function 'memcpy' [-Wimplicit-function-declaration]
   55 |       memcpy(buf, current_byte, realignment);
      |       ^~~~~~
sw/device/lib/base/mmio.c:10:1: note: include '<string.h>' or provide a declaration of 'memcpy'
    9 | #include "sw/device/lib/base/memory.h"
  +++ |+#include <string.h>
   10 |
sw/device/lib/base/mmio.c:55:7: warning: incompatible implicit declaration of built-in function 'memcpy' [-Wbuiltin-declaration-mismatch]
   55 |       memcpy(buf, current_byte, realignment);
      |       ^~~~~~
sw/device/lib/base/mmio.c:55:7: note: include '<string.h>' or provide a declaration of 'memcpy'
sw/device/lib/base/mmio.c:58:7: warning: incompatible implicit declaration of built-in function 'memcpy' [-Wbuiltin-declaration-mismatch]
   58 |       memcpy(current_byte, buf, realignment);
      |       ^~~~~~
sw/device/lib/base/mmio.c:58:7: note: include '<string.h>' or provide a declaration of 'memcpy'
sw/device/lib/base/mmio.c:89:7: warning: incompatible implicit declaration of built-in function 'memcpy' [-Wbuiltin-declaration-mismatch]
   89 |       memcpy(buf, &current_word, bytes_to_copy);
      |       ^~~~~~
sw/device/lib/base/mmio.c:89:7: note: include '<string.h>' or provide a declaration of 'memcpy'
sw/device/lib/base/mmio.c:92:7: warning: incompatible implicit declaration of built-in function 'memcpy' [-Wbuiltin-declaration-mismatch]
   92 |       memcpy(&current_word, buf, bytes_to_copy);
      |       ^~~~~~
sw/device/lib/base/mmio.c:92:7: note: include '<string.h>' or provide a declaration of 'memcpy'
```

--- a/sw/device/lib/base/memory.c
+++ b/sw/device/lib/base/memory.c
@@ -58,8 +58,7 @@ int OT_PREFIX_IF_NOT_RV32(memcmp)(const void *lhs, const void *rhs,
 }
 
 OT_WEAK
-int OT_PREFIX_IF_NOT_RV32(memrcmp)(const void *lhs, const void *rhs,
-                                   size_t len) {
+int memrcmp(const void *lhs, const void *rhs, size_t len) {
   const uint8_t *lhs8 = (uint8_t *)lhs;
   const uint8_t *rhs8 = (uint8_t *)rhs;
   size_t j;

--- a/sw/device/lib/base/memory.h
+++ b/sw/device/lib/base/memory.h
@@ -17,13 +17,23 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#ifndef OT_PLATFORM_RV32
+#include <string.h>
+#endif
+
 #include "sw/device/lib/base/macros.h"
 
 // When compiling unit tests on the host machine, we must mangle the names of
-// OpenTitan's memory functions to avoid conflicting with the libc variants. For
-// instance, without name mangling, the compiler may complain that `memcpy()`
-// has a different exception specifier than the variant declared in <string.h>.
+// OpenTitan's memory functions to disambiguate them from libc's variants.
+// Otherwise, the compiler could error when it encounters conflicting
+// declarations. For instance, OpenTitan's `memcpy()` could have a different
+// exception specifier than the one declared by the system header <string.h>.
 // (The <string.h> include is unavoidable because it's pulled in by GoogleTest.)
+//
+// Consumers of this header should almost always call the unmangled functions.
+// Device builds will naturally use OpenTitan's definitions and host builds will
+// use the system's definitions. However, nothing is stopping host builds from
+// calling the mangled function names; see :memory_unittest for an example.
 //
 // Separately, ASan segfaults during initialization when it calls our
 // user-defined memory functions. Prefixing the troublesome functions works
@@ -218,6 +228,9 @@ int OT_PREFIX_IF_NOT_RV32(memcmp)(const void *lhs, const void *rhs, size_t len);
  *
  * Can be used for arithmetic comparison of little-endian buffers.
  *
+ * This function is provided by OpenTitan in host builds, not by the platform's
+ * libc implementation.
+ *
  * @param lhs the left-hand-side of the comparison.
  * @param rhs the right-hand-side of the comparison.
  * @param len the length of both regions, in bytes.
@@ -225,8 +238,7 @@ int OT_PREFIX_IF_NOT_RV32(memcmp)(const void *lhs, const void *rhs, size_t len);
  * contingencies of `lhs == rhs`, `lhs > rhs`, and `lhs < rhs` (as buffers, not
  * pointers), respectively.
  */
-int OT_PREFIX_IF_NOT_RV32(memrcmp)(const void *lhs, const void *rhs,
-                                   size_t len);
+int memrcmp(const void *lhs, const void *rhs, size_t len);
 
 /**
  * Search a region of memory for the first occurence of a particular byte value.

--- a/sw/device/lib/base/memory_unittest.cc
+++ b/sw/device/lib/base/memory_unittest.cc
@@ -73,8 +73,8 @@ class MemChrTest : public ::testing::TestWithParam<decltype(&ot_memchr)> {};
 INSTANTIATE_TEST_SUITE_P(MemCpy, MemCpyTest,
                          ::testing::Values(ot_memcpy, __builtin_memcpy));
 INSTANTIATE_TEST_SUITE_P(MemCmp, MemCmpTest,
-                         ::testing::Values(ot_memcmp, __builtin_memcmp,
-                                           ot_memrcmp, ref_memrcmp));
+                         ::testing::Values(ot_memcmp, __builtin_memcmp, memrcmp,
+                                           ref_memrcmp));
 INSTANTIATE_TEST_SUITE_P(MemSet, MemSetTest,
                          ::testing::Values(ot_memset, __builtin_memset));
 INSTANTIATE_TEST_SUITE_P(MemChr, MemChrTest,
@@ -233,8 +233,7 @@ TEST_P(MemCmpTest, Properties) {
 TEST_P(MemCmpTest, DoesNotUseSystemEndianness) {
   auto memcmp_func = GetParam();
 
-  const bool reverse =
-      memcmp_func == &ot_memrcmp || memcmp_func == &ref_memrcmp;
+  const bool reverse = memcmp_func == &memrcmp || memcmp_func == &ref_memrcmp;
 
   constexpr uint8_t kBuf1[] = {0, 0, 0, 1};
   constexpr uint8_t kBuf2[] = {0, 0, 1, 0};

